diff --git a/node_modules/smelte/src/components/Tabs/Indicator.svelte b/node_modules/smelte/src/components/Tabs/Indicator.svelte
index f680542..1dee8dd 100644
--- a/node_modules/smelte/src/components/Tabs/Indicator.svelte
+++ b/node_modules/smelte/src/components/Tabs/Indicator.svelte
@@ -10,7 +10,7 @@
 </script>
 
 <div
-  class="absolute bottom-0 left-0 transition {bg(700)}"
+  class="absolute bottom-0 left-0 transition-all {bg(700)}"
   class:hidden={left < 0}
   style="width: {width}px; left: {left}px; height: 2px;"
-  transition:slide />
+  transition:slide|local />
